soc/nordic/common/dmm.h

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * nRF SoC specific public APIs for Device Memory Management (dmm) subsystem
+ */
+
+#ifndef SOC_NORDIC_COMMON_DMM_H_
+#define SOC_NORDIC_COMMON_DMM_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/** @cond INTERNAL_HIDDEN */
+
+/**
+ * @brief Prepare a DMA output buffer for the specified device
+ *
+ * Allocate an output buffer in memory region that given device can perform DMA transfers from.
+ * Copy @p user_buffer contents into it.
+ * Writeback data cache lines associated with output buffer, if needed.
+ *
+ * @note Depending on provided user buffer parameters and SoC architecture,
+ *       dynamic allocation and cache operations might be skipped.
+ *
+ * @warning It is prohibited to read or write @p user_buffer or @p buffer_out contents
+ *          until @ref dmm_dma_buffer_out_release() is called
+ *
+ * @param dev Device to prepare the buffer for.
+ * @param user_buffer Buffer containing data to be processed by the given device
+ * @param user_out Pointer to a pointer to be filled by prepared DMA buffer address.
+ *
+ * @retval 0 If succeeded.
+ * @retval -ENOMEM If output buffer could not be allocated.
+ * @retval -errno Negative errno for other failures.
+ */
+int dmm_dma_buffer_out_prepare(const struct device *dev, void const * user_buffer, void ** buffer_out);
+
+/**
+ * @brief Release the previously prepared DMA output buffer
+ *
+ * @param dev Device to release the buffer for.
```

Comment:
Is this the device that will use DMA (e.g. UARTE) or mempool device that is associated with the DMA device through DT reference?
Imo, it cannot be the peripheral because there is no generic, runtime device API to get mempool reference from the device.

Reply:
No. The device drivers must be generic: they must work correctly with the compatible peripherals regardless of how they are integrated into the system (which memory is DMAble by them, what are properties of this memory). The integration information should be included in the device tree properties of the compatible node and retrieved by the device driver. The device driver should use only the node it instantiates.

If a device tree node misses a property like `mempool`, then maybe the driver should not use `dmm` at all.

Reply:
Scratch my previous message. I misunderstood the proposal.

A device driver X uses a reference to a device tree node it instantiates (N), with no changes compared to the current device drivers.
What we would modify is to add a property M to node N being a reference to a mempool. The device driver X would retrieve property M from node N and use `const struct dev*` (D) represented by M. When X calls `dmm` functions it would use D as the `dev` argument.

If I correctly described the proposal, then it's feasible.

Reply:
I see another issue here. In `dmm.c` `_prepare` API we need to check whether the provided buffer is already in correct memory region and if not, use mempool device to dynamically allocate new buffer.

Does it mean we actually need to provide **two** devices to `_prepare` API:
- reference to RAM region node, associated with specified peripheral node (to check buffer location),
- reference to mempool device (to perform dynamic allocation if buffer is in incorrect location) ?

If yes, I am wondering if these two could be somehow linked, so we provide single device instead of two.

> Imo, it cannot be the peripheral because there is no generic, runtime device API to get mempool reference from the device.
@nordic-krch would be it feasible to add such runtime device API? I believe it would improve the abstraction and also solve this issue

Reply:
> Does it mean we actually need to provide **two** devices to `_prepare` API

Good point. I don't think a device tree node should have an indication of software configuration like if a given memory region contains a pool for dynamically allocated memory. The device tree node should describe the hardware properties like which memory region is DMA-able by this peripheral. Something like:

```
ram30: memory@0x... {
  properties = NON_CACHEABLE;
};

uart130 {
  compatible = "nordic,...";
  memory = <&ram30 &ram31>;
};
```

And dmm should figure out internally which allocator to use to correctly allocate a buffer in `memory` pointed by `&uart130`.

Reply:
updated PR - now `dmm` accepts memory region address pointer as first parameter. Device drivers (and other components) can easily obtain this value by peeking at devicetree property. `dmm` has internal list of regions and finds the one matching the one provided by the user of the API